src/box/lua/net_box.c

```diff
@@ -2114,40 +2374,181 @@ netbox_iproto_schema(struct lua_State *L)
-	struct netbox_transport *transport = luaT_check_netbox_transport(L, 2);
-	int fd = lua_tointeger(L, 3);
+	struct lua_State *L = transport->L;
+	transport->state = NETBOX_ACTIVE;
+	netbox_transport_on_state_change(transport, L);
 	while (true) {
 		struct xrow_header hdr;
-		if (netbox_send_and_recv(transport, fd, &hdr) != 0)
-			return luaT_error(L);
-		netbox_dispatch_response(L, transport, &hdr);
+		if (netbox_transport_send_and_recv(transport, &hdr) != 0)
+			luaT_error(L);
+		netbox_transport_dispatch_response(transport, &hdr);
 		if (hdr.schema_version > 0 &&
 		    hdr.schema_version != schema_version) {
-			return 0;
+			break;
 		}
 	}
 }
 
+/**
+ * Connection handler. Raises a Lua error on termination.
+ */
+static int
+netbox_connection_handler_f(struct lua_State *L)
+{
+	struct netbox_transport *transport = (void *)lua_topointer(L, 1);
+	netbox_transport_do_id(transport);
+	netbox_transport_do_auth(transport);
+	while (true) {
+		uint32_t schema_version;
+		schema_version = netbox_transport_fetch_schema(transport);
+		netbox_transport_process_requests(transport, schema_version);
+	}
+	return 0;
+}
+/**
+ * Worker fiber routine.
+ */
+static int
+netbox_worker_f(va_list ap)
+{
+	struct netbox_transport *transport;
+	transport = va_arg(ap, struct netbox_transport *);
+	assert(transport->worker == fiber());
+	assert(transport->L != NULL);
+	assert(transport->coro_ref != LUA_NOREF);
+	assert(transport->self_ref != LUA_NOREF);
+	struct lua_State *L = transport->L;
+	const double reconnect_after = transport->opts.reconnect_after;
+	while (!fiber_is_cancelled()) {
+		if (netbox_transport_connect(transport) == 0) {
+			int rc = luaT_cpcall(L, netbox_connection_handler_f,
+					     transport);
+			/* The worker loop can only be broken by an error. */
+			assert(rc != 0);
+			(void)rc;
+			iostream_close(&transport->io);
+		}
+		if (transport->state == NETBOX_CLOSED)
+			break;
+		netbox_transport_set_error(transport);
+		transport->state = (reconnect_after > 0 ?
+				    NETBOX_ERROR_RECONNECT : NETBOX_ERROR);
+		netbox_transport_on_state_change_pcall(transport, L);
+		if (reconnect_after > 0) {
+			fiber_sleep(reconnect_after);
+		} else {
+			break;
+		}
+	}
+	transport->worker = NULL;
+	transport->L = NULL;
+	luaL_unref(tarantool_L, LUA_REGISTRYINDEX, transport->coro_ref);
+	transport->coro_ref = LUA_NOREF;
+	/* Careful: luaL_unref may delete this transport object. */
+	int ref = transport->self_ref;
+	transport->self_ref = LUA_NOREF;
+	luaL_unref(tarantool_L, LUA_REGISTRYINDEX, ref);
+	return 0;
+}
+
+/**
+ * Starts a worker fiber.
+ */
+static int
+luaT_netbox_transport_start(struct lua_State *L)
+{
+	struct netbox_transport *transport = luaT_check_netbox_transport(L, 1);
+	assert(transport->worker == NULL);
+	assert(transport->L == NULL);
+	assert(transport->coro_ref == LUA_NOREF);
+	assert(transport->self_ref = LUA_NOREF);
+	transport->L = luaT_newthread(L);
+	if (transport->L == NULL)
+		return luaT_error(L);
```

Comment:
If you throw the error anyway, you can use `lua_newthread` as is.

Reply:
Right. Fixed. Thanks!